grafana/grafana
# imagestorage: fix templating of public_url option for webdav

This Pull Request (PR) addresses a long-standing bug (#15213) related to WebDAV public URL formatting in Grafana's image upload configuration.

## Bug description

In Grafana, there are two URL configuration options for uploading images: `url` and `public_url`. The `url` option specifies the upload destination, while `public_url` is used to format the link to the uploaded image in alert notifiers, such as emails or Slack messages. The issue arises when users attempt to template the `public_url` by adding `${file}` to dynamically insert the image filename. The bug causes the templating to fail, resulting in an incorrect URL format.

The user configures `public_url` like this:

 `public_url = https://example.com/files?p=/${file}`

and the result is (if the filename is 123456.png):

`https://example.com/files/123456.png?p=/`

The expected result is:

`https://example.com/files?p=/123456.png`

## Reason

The bug occurs during the configuration file loading process when Grafana mistakenly removes the `${file}` template string. The bug was introduced by the Expander code that interpolates environment variables in the config file (#25075). This happens because there is no environment variable named `file`, and during the configuration loading phase, `${file}` gets overwritten with an empty string. 

## Recreating the bug

This is a difficult bug to recreate which is probably why it has never been fixed. In this PR, I add a docker block for WebDAV storage. If you don't have a WebDAV server then Grafana returns an error while uploading the image and you don't see the `public_url` bug.

I created a simple alerting webhook implementation to be able to see the image link in the alert notifier message: https://github.com/daniellee/grafana-webhook. But you can see the bug by debugging or checking an alert email as well.

To recreate the bug follow these steps:

1. Configure the external image storage and the WebDAV option in the config file:
   ```
   [external_image_storage]
   provider = webdav
   
   [external_image_storage.webdav]
   url = http://127.0.0.1:5000/images
   public_url = http://127.0.0.1:5000/images?p=/${file}

   username = test
   password = test
   ```
1. Configure unified alerting:
    ```
    [unified_alerting]
    enabled = true

    [unified_alerting.screenshots]
    capture = true
    upload_external_image_storage = true
    ```
1. Setup Webdav storage. This is a simple WebDAV server in demo mode with no users or passwords that runs on http://localhost:5000 (or use the docker-compose file in this PR):
    ```
    docker run -v `pwd`:/data -p 5000:5000 --rm -it sigoden/dufs /data -A
    ```
1. Create a Grafana alert that fires and a notification channel (like email) that uses images in the alert notification (so not a Mimir / Cortex / Loki alert as they don't have images in alerts).
1. In the email, the image link will be broken (as the WebDAV server is running locally and email doesn't have access to it). But you can check the link to see if it is correctly formatted.

## Proposed solution

This PR proposes changing the template syntax to be `{{file}}` instead of `${file}`. This doesn't get expanded and accidentally removed when the config file is loaded. The new syntax is consistent with Go templating used for alert notifications.

```
[external_image_storage.webdav]
url = https://example.com/Public/prometheus
public_url = https://example.com//files/?p=/{{file}}
```

While this may be considered a breaking change, it hasn't worked as intended for years so it won't break anything for any users.

Fixes #15213
